composer/algorithms/mixup/mixup.py

```diff
@@ -127,46 +118,67 @@ class MixUp(Algorithm):
+        else:
+            return event in [Event.AFTER_DATALOADER, Event.BEFORE_LOSS]
 
     def apply(self, event: Event, state: State, logger: Logger) -> None:
-        """Applies MixUp augmentation on State input.
-
-        Args:
-            event (Event): the current event
-            state (State): the current trainer state
-            logger (Logger): the training logger
-        """
-
-        input, target = state.batch_pair
-        assert isinstance(input, torch.Tensor) and isinstance(target, torch.Tensor), \
-            "Multiple tensors for inputs or targets not supported yet."
-
-        self.mixing = _gen_mixing_coef(self.alpha)
-
-        new_input, new_target, self.indices = mixup_batch(
-            input,
-            target,
-            mixing=self.mixing,
-            num_classes=self.num_classes,
-        )
+        if event == Event.INIT:
+            # Grab the loss function
+            if isinstance(state.model, ComposerModel):
+                self._loss_fn = state.model.loss
+            else:
+                raise RuntimeError("Model must be of type ComposerModel")
+
+        if event == Event.AFTER_DATALOADER:
+            input, target = state.batch_pair
+            assert isinstance(input, torch.Tensor) and isinstance(target, torch.Tensor), \
+                "Multiple tensors for inputs or targets not supported yet."
+
+            self.mixing = _gen_mixing_coef(self.alpha)
+            self.indices = _gen_indices(input.shape[0])
+
+            new_input, _, self.permuted_target, _ = mixup_batch(
+                input,
+                target,
+                mixing=self.mixing,
+                indices=self.indices,
+            )
 
-        state.batch = (new_input, new_target)
+            state.batch = (new_input, target)
+
+        if not self.interpolate_loss and event == Event.BEFORE_LOSS:
+            # Interpolate the targets
+            input, target = state.batch_pair
+            assert isinstance(state.outputs, torch.Tensor), "Multiple output tensors not supported yet"
+            assert isinstance(target, torch.Tensor), "Multiple target tensors not supported yet"
+            # Make sure that the targets are dense/one-hot
+            target = ensure_targets_one_hot(state.outputs, target)
+            permuted_target = ensure_targets_one_hot(state.outputs, self.permuted_target)
+            # Interpolate to get the new target
+            mixed_up_target = (1 - self.mixing) * target + self.mixing * permuted_target
+            # Create the new batch
+            state.batch = (input, mixed_up_target)
+
+        if self.interpolate_loss and event == Event.AFTER_LOSS:
+            input, target = state.batch_pair
+            # Interpolate the loss
+            assert self._loss_fn is not None, "loss_fn should be set on Event.INIT"
+            new_loss = self._loss_fn(state.outputs, (torch.Tensor(), self.permuted_target))
```

Comment:
Do we want snapshotted loss from INIT here, or do we want to read whatever the loss function currently is at this point? Thinking about case of someone combining several methods that alter the loss function, with mixup not last in the order. The snapshot-ed loss might just the vanilla xent, but the current loss could be something more complex. Using only the former will make the permuted targets use a different loss than the original targets, which is no longer equivalent to interpolating the targets during AFTER_BATCH.

Reply:
also I'm a little confused as to why we construct an empty Tensor here in the call to `self._loss_fn`. Is it just to get it to typecheck? This probably works for most loss functions, but lying to an arbitrary user `ComposerModel` about the batch input scares me--the user's code might assume it's getting the real input, or at least a tensor of the correct shape. Could we just pass in the input tensor?

Reply:
Switched to passing the actual input to the loss instead of an empty tensor. 

For grabbing the loss on INIT, this is mostly to avoid having to do extra checks on `state.model`, since after INIT it will sometimes be wrapped by either torch ddp or deepspeed, in which case `state.model.loss` will no longer exist. In the ddp case, things can still work if there's an explicit check for ddp, and `state.model.module.loss` is used instead. Unsure about the deepspeed case. I agree that this can/will cause problems down the line for composition, so perhaps it's worth handling each case explicitly. This discussion is also relevant for [selective backprop](https://github.com/mosaicml/composer/blob/4bcda6e50474df4073e2e69e3443bb9980482239/composer/algorithms/selective_backprop/selective_backprop.py#L222) which does something similar. What do you think?

Reply:
hmm...you're right, that's tricky. @jbloxham, when using deepspeed, how would one access the ComposerModel?

I'm leaning towards just `try...excepting` access to `state.model.loss`, `state.module.module.loss`, and whatever deepspeed does if necessary. IMO not supporting composition with this right now is taking on tech debt that we'll have to refactor or work around at some point.

Reply:
Thinking about this a bit more, I don't think it's currently possible for the training loss function to be different from `state.model.loss` at INIT since the loss that actually gets computed by the trainer comes from `self.original_model` which is identical to `state.model` at INIT, and cannot be changed by algorithms. So, the function mixup grabs will always be the one used to compute `state.loss` unless something changes with the trainer. There can be composition issues if other algorithms try to modify the loss the way mixup does here, but that should be able to be handled by setting the proper algorithm order [here](https://github.com/mosaicml/composer/blob/4bcda6e50474df4073e2e69e3443bb9980482239/composer/core/engine.py#L231) when possible (e.g. adding a label smoothing term after mixup), or raising the appropriate error when composition cannot work as intended (e.g. cutmix + mixup won't compose correctly in any order if cutmix were also implemented this way).

So, as is, things should be ok, but the trainer should probably be changed such that it is possible for algorithms to modify the loss.  I think we should use the more future-proof way of getting the loss function right before the mixup loss is calculated, if it's possible with deepspeed.

Reply:
Pretty sure `state.model.module.loss` also works for DeepSpeed.

Also note this PRD, which would provide a much more direct way of accessing the loss function if implemented: https://www.notion.so/457ebaced580405c9e87eb0032f034f7?v=ff1ae88e75294fcaa796d004d7c06504&p=c4c247ee219a4e238b50ee4cca8201f4

Reply:
Ok, made this change.